### ISSUE TITLE
Support for GitLab pipelines

### ISSUE THREAD

**Issue by lachmanfrantisek** [packit/ogr#420]
Pipelines are specific to the GitLab but there were some requests (ping @thrix ) to support them in OGR as well.

- It's going against the purpose of OGR to have one API for multiple forges. :-1: 
- Users can access `python-gitlab` to do that instead. :-1: 
  - But it's not a good UX. `python-gitlab` is our implementation detail. We should not force users to access that. :+1: 
- Technically it will be easy to implement. :+1: (No big changes needed, just add some GitLab specific classes and methods.)
- It will be useful for packit-service GitLab implementation :+1: (ping @shreyaspapi ..;-)

AC:
- [ ] Provide a class abstraction on top of the objects related to that.
  - Make here a proposal before coding.
- [ ] Implement useful methods and classes.
- [ ] Create new issues for anything that can be implemented later.

**Comment by csomh** [comment]
Could users use `python-gitlab` directly? How does the user experience look like, when one uses `ogr` for things common to all forges and `python-gitlab` to do GitLab pipeline specific things?

What I'm trying to understand is whether there is a point in wrapping `python-gitlab`'s pipeline support into `ogr`. What are the benefits compared to just using `python-gitlab` when it comes to pipelines?

**Comment by lachmanfrantisek** [comment]
> Could users use python-gitlab directly? 

Yes, as I wrote in the description. (We usually have some private attribute of our class containing the `gitlab-python` object.)

> How does the user experience look like, when one uses ogr for things common to all forges and python-gitlab to do GitLab pipeline specific things?

Just a few forge-specific classes/methods. Nothing that will break our architecture. We already have some forge-specific methods. (This is more a philosophical problem than the technical one...;)

> What I'm trying to understand is whether there is a point in wrapping python-gitlab's pipeline support into ogr. What are the benefits compared to just using python-gitlab when it comes to pipelines?

- Two different approaches, worse code quality and readability
  - `gitlab-python` sometimes don't specify arguments which can be used; listing objects is a bit specific and not documented -- @mfocko can say more here..;) #392 
  - Editing remote objects needs to call the save methods to actually save the change, we (=OGR) handle the saving for you.
- We will probably use it when supporting the GitLab in packit-service (This is exactly what we are missing on GitHub, some display of the workflow steps.) => It makes sense to ease our lives.
- GitLab methods/functions/classes are not (obviously) connected to the OGR ones -> You cannot go back from the `gitlab-python` pipeline objects to the OGR representation of PR, project, service (and commit/branch/.. in the future).

**Comment by csomh** [comment]
> * GitLab methods/functions/classes are not (obviously) connected to the OGR ones -> You cannot go back from the `gitlab-python` pipeline objects to the OGR representation of PR, project, service (and commit/branch/.. in the future).

:+1: To me this is the most obvious benefit for having GitLab Pipeline support in OGR. Thanks for talking through this!

**Comment by lachmanfrantisek** [comment]
I would still like to see it implemented. At least basic object that will contain `gitlab-python` representation as an attribute.

**Comment by lachmanfrantisek** [comment]
We will need this soon...

**Comment by lachmanfrantisek** [comment]
Since we already have "check runs" for GitHub, we can add this for GitLab in a similar way.
